File: desktop/windows/src/renderer/src/device/session/deviceOperationBroker.ts
/**
 * Device operation broker and uncorrelated-callback gate — Windows port of
 * macOS Session/DeviceOperationBroker.swift.
 *
 * The broker serializes correlated request/response exchanges: at most one
 * pending operation per key, an optional timeout on the injected clock, and
 * exactly one terminal event resuming the caller (late timeout/cancel/succeed
 * events are explicit no-ops; cancelAll supersedes even claimed completions).
 *
 * The gate exists for operations whose physical callback identifies only a
 * key (a characteristic or command id), not the attempt. Once an operation
 * terminates without success, a late callback for that key can never again be
 * attributed to a newer operation, so the key is poisoned until session reset.
 */

export type DeviceOperationTermination =
  | 'succeeded'
  | 'timedOut'
  | 'cancelled'
  | 'disconnected'
  | 'failed'

export type DeviceOperationBrokerErrorKind =
  | 'operationAlreadyPending'
  | 'timedOut'
  | 'cancelled'
  | 'disconnected'
  | 'failed'

const BROKER_ERROR_MESSAGES: Record<Exclude<DeviceOperationBrokerErrorKind, 'failed'>, string> = {
  operationAlreadyPending: 'Another operation with the same identity is already pending',
  timedOut: 'Operation timed out',
  cancelled: 'Operation was cancelled',
  disconnected: 'Device disconnected before the operation completed'
}

export class DeviceOperationBrokerError extends Error {
  readonly kind: DeviceOperationBrokerErrorKind

  constructor(kind: DeviceOperationBrokerErrorKind, detail?: string) {
    super(kind === 'failed' ? (detail ?? 'Operation failed') : BROKER_ERROR_MESSAGES[kind])
    this.name = 'DeviceOperationBrokerError'
    this.kind = kind
  }
}

/** Identity of one gate-registered attempt: key plus a monotonic token. */
export interface DeviceOperationHandle {
  key: string
  token: number
}

// --- clock ------------------------------------------------------------------

/** Injectable clock: every sleep/timeout in the device stack goes through
 *  this seam so tests can drive time deterministically. */
export interface DeviceOperationClock {
  sleep(ms: number, signal: AbortSignal): Promise<'elapsed' | 'aborted'>
}

export class RealDeviceOperationClock implements DeviceOperationClock {
  sleep(ms: number, signal: AbortSignal): Promise<'elapsed' | 'aborted'> {
    return new Promise((resolve) => {
      if (signal.aborted) {
        resolve('aborted')
        return
      }
      const timer = setTimeout(() => {
        signal.removeEventListener('abort', onAbort)
        resolve('elapsed')
      }, ms)
      const onAbort = (): void => {
        clearTimeout(timer)
        resolve('aborted')
      }
      signal.addEventListener('abort', onAbort, { once: true })
    })
  }
}

// --- uncorrelated-callback gate ---------------------------------------------

export class UncorrelatedOperationGate {
  private live = new Map<string, DeviceOperationHandle>()
  private poisoned = new Set<string>()
  private nextToken = 1

  /** A key can host a new attempt only when nothing is in flight and no prior
   *  attempt left an uncorrelated callback behind. */
  canStart(key: string): boolean {
    return !this.live.has(key) && !this.poisoned.has(key)
  }

  isPoisoned(key: string): boolean {
    return this.poisoned.has(key)
  }

  /** True while an attempt is registered and unclaimed on the key. */
  hasLiveAttempt(key: string): boolean {
    return this.live.has(key)
  }

  /** Registers an attempt; null when the key is unavailable (in flight or
   *  poisoned) — callers map that to their own identity error message. */
  register(key: string): DeviceOperationHandle | null {
    if (!this.canStart(key)) return null
    const handle: DeviceOperationHandle = { key, token: this.nextToken++ }
    this.live.set(key, handle)
    return handle
  }

  /** Claims the live attempt for an arriving physical callback. Null means
   *  the callback is uncorrelated (no attempt, or the key is poisoned) and
   *  must not be treated as a response. */
  takeHandleForCallback(key: string): DeviceOperationHandle | null {
    if (this.poisoned.has(key)) return null
    const handle = this.live.get(key)
    if (handle === undefined) return null
    this.live.delete(key)
    return handle
  }

  /** Records the attempt's outcome. Any non-success termination poisons the
   *  key permanently until reset(): the physical callback may still arrive
   *  later and must never be attributed to a newer operation. */
  terminal(handle: DeviceOperationHandle, termination: DeviceOperationTermination): void {
    const current = this.live.get(handle.key)
    if (current !== undefined && current.token === handle.token) {
      this.live.delete(handle.key)
    }
    if (termination !== 'succeeded') {
      this.poisoned.add(handle.key)
    }
  }

  /** Session teardown: a new physical connection gets clean identities. */
  reset(): void {
    this.live.clear()
    this.poisoned.clear()
  }
}

// --- broker -----------------------------------------------------------------

interface PendingOperation {
  key: string
  settled: boolean
  claimed: boolean
  startTask: Promise<void>
  startFailed: boolean
  startError: unknown
  timeoutAbort: AbortController | null
  onTerminal: ((termination: DeviceOperationTermination) => void) | null
  resolve: (value: unknown) => void
  reject: (error: unknown) => void
}

export interface PerformOptions {
  key: string
  /** Milliseconds; omit or null for no timeout. */
  timeoutMs?: number | null
  /** Fires with the mapped termination before the caller's promise settles. */
  onTerminal?: (termination: DeviceOperationTermination) => void
  /** Kicks off the physical operation. A start failure is surfaced when the
   *  operation is later completed via succeed(). */
  start: () => void | Promise<void>
}

export class DeviceOperationBroker {
  private pending = new Map<string, PendingOperation>()

  constructor(private readonly clock: DeviceOperationClock = new RealDeviceOperationClock()) {}

  hasPending(key: string): boolean {
    return this.pending.has(key)
  }

  perform<T>(options: PerformOptions): Promise<T> {
    if (this.pending.has(options.key)) {
      return Promise.reject(new DeviceOperationBrokerError('operationAlreadyPending'))
    }

    let resolve!: (value: unknown) => void
    let reject!: (error: unknown) => void
    const promise = new Promise<unknown>((res, rej) => {
      resolve = res
      reject = rej
    })

    const op: PendingOperation = {
      key: options.key,
      settled: false,
      claimed: false,
      startTask: Promise.resolve(),
      startFailed: false,
      startError: null,
      timeoutAbort: null,
      onTerminal: options.onTerminal ?? null,
      resolve,
      reject
    }
    this.pending.set(options.key, op)

    op.startTask = (async () => {
      try {
        await options.start()
      } catch (error) {
        op.startFailed = true
        op.startError = error
        // Settle now rather than waiting for a completion that a failed start
        // will never produce: without a timeout the caller would hang forever,
        // and with one it would report a misleading timeout.
        if (!op.settled && !op.claimed) {
          this.settle(op, 'failed', { reject: error })
        }
      }
    })()

    const timeoutMs = options.timeoutMs ?? null
    if (timeoutMs !== null) {
      const abort = new AbortController()
      op.timeoutAbort = abort
      void this.clock.sleep(timeoutMs, abort.signal).then((outcome) => {
        if (outcome !== 'elapsed') return
        if (op.settled || op.claimed) return
        this.settle(op, 'timedOut', {
          reject: new DeviceOperationBrokerError('timedOut')
        })
      })
    }

    return promise as Promise<T>
  }

  /** Completes the pending operation for a key with a value. The claim is
   *  exclusive against timeout/fail, but cancelAll supersedes it while the
   *  start task is still being awaited. A failed start throws its own error
   *  instead of returning the value. */
  succeed(key: string, value: unknown): void {
    const op = this.pending.get(key)
    if (op === undefined || op.settled || op.claimed) return
    op.claimed = true
    void (async () => {
      await op.startTask
      if (op.settled) return
      if (op.startFailed) {
        this.settle(op, 'failed', { reject: op.startError })
        return
      }
      this.settle(op, 'succeeded', { resolve: value })
    })()
  }

  /** Fails the pending operation for a key (e.g. an error-bearing callback). */
  fail(key: string, error: Error): void {
    const op = this.pending.get(key)
    if (op === undefined || op.settled || op.claimed) return
    this.settle(op, 'failed', { reject: error })
  }

  /** Terminates every pending operation, superseding even claimed completions
   *  whose start task is still settling. */
  cancelAll(reason: 'cancelled' | 'disconnected' = 'cancelled'): void {
    const ops = Array.from(this.pending.values())
    for (const op of ops) {
      if (op.settled) continue
      this.settle(op, reason, { reject: new DeviceOperationBrokerError(reason) })
    }
  }

  private settle(
    op: PendingOperation,
    termination: DeviceOperationTermination,
    outcome: { resolve?: unknown; reject?: unknown }
  ): void {
    if (op.settled) return
    op.settled = true
    op.timeoutAbort?.abort()
    if (this.pending.get(op.key) === op) {
      this.pending.delete(op.key)
    }
    try {
      op.onTerminal?.(termination)
    } catch (error) {
      // A terminal hook decides nothing about the caller: rethrowing here would
      // both strand the promise and surface as an unhandled rejection.
      console.warn('[device] operation terminal hook threw:', error)
    }
    if ('resolve' in outcome) {
      op.resolve(outcome.resolve)
    } else {
      op.reject(outcome.reject)
    }
  }
}

File: desktop/macos/Desktop/Sources/Bluetooth/Session/DeviceOperationBroker.swift
import Foundation

enum DeviceOperationBrokerError: LocalizedError, Equatable, Sendable {
  case operationAlreadyPending
  case timedOut
  case cancelled
  case disconnected
  case failed(String)

  var errorDescription: String? {
    switch self {
    case .operationAlreadyPending:
      return "Another operation with the same identity is already pending"
    case .timedOut:
      return "Operation timed out"
    case .cancelled:
      return "Operation was cancelled"
    case .disconnected:
      return "Device disconnected before the operation completed"
    case .failed(let reason):
      return reason
    }
  }
}

struct DeviceOperationHandle<Key: Hashable & Sendable>: Hashable, Sendable {
  fileprivate let key: Key
  fileprivate let token: UInt64
}

enum DeviceOperationTermination: Equatable, Sendable {
  case succeeded
  case timedOut
  case cancelled
  case disconnected
  case failed
}

/// Tracks operations whose physical callback identifies only a key, not the
/// broker token. If such an operation terminates before its callback arrives,
/// the key is poisoned until session teardown: starting another operation
/// would let the old callback masquerade as the new operation's response.
@MainActor
final class UncorrelatedOperationGate<Key: Hashable & Sendable> {
  private var handles: [Key: DeviceOperationHandle<Key>] = [:]
  private var poisonedKeys: Set<Key> = []

  func canStart(_ key: Key) -> Bool {
    handles[key] == nil && !poisonedKeys.contains(key)
  }

  @discardableResult
  func register(_ handle: DeviceOperationHandle<Key>) -> Bool {
    guard canStart(handle.key) else { return false }
    handles[handle.key] = handle
    return true
  }

  func takeHandleForCallback(key: Key) -> DeviceOperationHandle<Key>? {
    guard !poisonedKeys.contains(key) else { return nil }
    return handles.removeValue(forKey: key)
  }

  func terminal(
    handle: DeviceOperationHandle<Key>,
    termination: DeviceOperationTermination
  ) {
    guard handles[handle.key] == handle else { return }
    handles.removeValue(forKey: handle.key)
    if termination != .succeeded {
      poisonedKeys.insert(handle.key)
    }
  }

  func reset() {
    handles.removeAll()
    poisonedKeys.removeAll()
  }
}

protocol DeviceOperationClock: Sendable {
  func sleep(for duration: Duration) async throws
}

struct ContinuousDeviceOperationClock: DeviceOperationClock {
  func sleep(for duration: Duration) async throws {
    try await ContinuousClock().sleep(for: duration)
  }
}

/// Owns one pending operation per key and guarantees exactly one terminal result.
///
/// Device callbacks, timeouts, cancellation, and disconnect can all race. They
/// converge here so only the first terminal event removes and resumes a pending
/// continuation; every late event is an explicit no-op.
actor DeviceOperationBroker<Key: Hashable & Sendable, Value: Sendable> {
  private struct PendingOperation {
    let handle: DeviceOperationHandle<Key>
    let continuation: CheckedContinuation<Value, Error>
    let onTerminal:
      @MainActor @Sendable (
        DeviceOperationHandle<Key>,
        DeviceOperationTermination
      ) -> Void
    var timeoutTask: Task<Void, Never>?
    var startTask: Task<Result<Void, DeviceOperationBrokerError>, Never>?
    var completionToken: UInt64?
  }

  private struct CompletionClaim {
    let operation: PendingOperation
    let token: UInt64
  }

  private let clock: any DeviceOperationClock
  private var nextToken: UInt64 = 0
  private var nextCompletionToken: UInt64 = 0
  private var pending: [Key: PendingOperation] = [:]

  init(clock: any DeviceOperationClock = ContinuousDeviceOperationClock()) {
    self.clock = clock
  }

  func perform(
    key: Key,
    timeout: Duration? = nil,
    onTerminal:
      @escaping @MainActor @Sendable (
        DeviceOperationHandle<Key>,
        DeviceOperationTermination
      ) -> Void = { _, _ in },
    start: @escaping @MainActor @Sendable (DeviceOperationHandle<Key>) async throws -> Void
  ) async throws -> Value {
    guard pending[key] == nil else {
      throw DeviceOperationBrokerError.operationAlreadyPending
    }
    guard !Task.isCancelled else {
      throw DeviceOperationBrokerError.cancelled
    }

    nextToken &+= 1
    let handle = DeviceOperationHandle(key: key, token: nextToken)

    return try await withTaskCancellationHandler {
      try await withCheckedThrowingContinuation { continuation in
        var operation = PendingOperation(
          handle: handle,
          continuation: continuation,
          onTerminal: onTerminal,
          timeoutTask: nil,
          startTask: nil,
          completionToken: nil
        )

        if let timeout {
          operation.timeoutTask = Task { [clock] in
            do {
              try await clock.sleep(for: timeout)
            } catch {
              return
            }
            await self.finish(
              handle: handle,
              result: .failure(DeviceOperationBrokerError.timedOut)
            )
          }
        }

        pending[key] = operation
        if Task.isCancelled {
          Task {
            await self.finish(
              handle: handle,
              result: .failure(DeviceOperationBrokerError.cancelled)
            )
          }
          return
        }
        let startTask = Task<Result<Void, DeviceOperationBrokerError>, Never> { @MainActor in
          guard !Task.isCancelled,
            await self.isPending(handle: handle)
          else {
            return Result.failure(DeviceOperationBrokerError.cancelled)
          }
          do {
            try await start(handle)
            return Result.success(())
          } catch {
            return Result.failure(
              DeviceOperationBrokerError.failed(error.localizedDescription)
            )
          }
        }
        pending[key]?.startTask = startTask
        Task {
          guard case .failure(let error) = await startTask.value else { return }
          await self.finish(handle: handle, result: .failure(error))
        }
      }
    } onCancel: {
      Task {
        await self.finish(
          handle: handle,
          result: .failure(DeviceOperationBrokerError.cancelled)
        )
      }
    }
  }

  @discardableResult
  func succeed(handle: DeviceOperationHandle<Key>, value: Value) async -> Bool {
    guard let claim = claimPending(handle: handle) else { return false }
    let operation = claim.operation
    operation.timeoutTask?.cancel()
    let startResult = await operation.startTask?.value
    guard removePending(handle: handle, completionToken: claim.token) != nil else {
      return false
    }
    if let startResult, case .failure(let error) = startResult {
      await operation.onTerminal(handle, termination(for: error))
      operation.continuation.resume(throwing: error)
    } else {
      await operation.onTerminal(handle, .succeeded)
      operation.continuation.resume(returning: value)
    }
    return true
  }

  @discardableResult
  func fail(handle: DeviceOperationHandle<Key>, reason: String) async -> Bool {
    guard let claim = claimPending(handle: handle) else { return false }
    let operation = claim.operation
    operation.timeoutTask?.cancel()
    if let startTask = operation.startTask {
      _ = await startTask.value
    }
    guard removePending(handle: handle, completionToken: claim.token) != nil else {
      return false
    }
    await operation.onTerminal(handle, .failed)
    operation.continuation.resume(
      throwing: DeviceOperationBrokerError.failed(reason)
    )
    return true
  }

  func cancelAll(reason: DeviceOperationBrokerError = .cancelled) async {
    // Session shutdown deliberately supersedes even a callback that has
    // already claimed completion. Claim every current entry before the
    // first suspension so keys remain unavailable while physical starts
    // drain, and so only this cancellation wave can resume them.
    let handles = pending.values.map(\.handle)
    let claims = handles.compactMap { handle in
      claimPending(handle: handle, replacingExistingClaim: true)
    }
    for claim in claims {
      let operation = claim.operation
      operation.timeoutTask?.cancel()
      operation.startTask?.cancel()
      if let startTask = operation.startTask {
        _ = await startTask.value
      }
      guard
        removePending(
          handle: operation.handle,
          completionToken: claim.token
        ) != nil
      else {
        continue
      }
      await operation.onTerminal(operation.handle, termination(for: reason))
      operation.continuation.resume(throwing: reason)
    }
  }

  func hasPendingOperation(for key: Key) -> Bool {
    pending[key] != nil
  }

  var pendingCount: Int {
    pending.count
  }

  private func finish(
    handle: DeviceOperationHandle<Key>,
    result: Result<Value, Error>
  ) async {
    // Ordinary terminal sources (timeout, caller cancellation, and start
    // failure) may only claim an unclaimed operation. Once a physical
    // callback owns completion, they are late signals and must not steal
    // or double-resume the continuation while its start task drains.
    guard let claim = claimPending(handle: handle) else { return }
    let operation = claim.operation
    operation.timeoutTask?.cancel()
    operation.startTask?.cancel()
    if let startTask = operation.startTask {
      _ = await startTask.value
    }
    guard removePending(handle: handle, completionToken: claim.token) != nil else {
      return
    }
    await operation.onTerminal(handle, termination(for: result))
    operation.continuation.resume(with: result)
  }

  private func isPending(handle: DeviceOperationHandle<Key>) -> Bool {
    pending[handle.key]?.handle == handle
  }

  private func removePending(
    handle: DeviceOperationHandle<Key>,
    completionToken: UInt64
  ) -> PendingOperation? {
    guard pending[handle.key]?.handle == handle,
      pending[handle.key]?.completionToken == completionToken
    else {
      return nil
    }
    return pending.removeValue(forKey: handle.key)
  }

  private func claimPending(
    handle: DeviceOperationHandle<Key>,
    replacingExistingClaim: Bool = false
  ) -> CompletionClaim? {
    guard var operation = pending[handle.key],
      operation.handle == handle,
      replacingExistingClaim || operation.completionToken == nil
    else {
      return nil
    }
    nextCompletionToken &+= 1
    operation.completionToken = nextCompletionToken
    pending[handle.key] = operation
    return CompletionClaim(operation: operation, token: nextCompletionToken)
  }

  private func termination(
    for result: Result<Value, Error>
  ) -> DeviceOperationTermination {
    switch result {
    case .success:
      return .succeeded
    case .failure(let error as DeviceOperationBrokerError):
      return termination(for: error)
    case .failure:
      return .failed
    }
  }

  private func termination(
    for error: DeviceOperationBrokerError
  ) -> DeviceOperationTermination {
    switch error {
    case .timedOut:
      return .timedOut
    case .cancelled:
      return .cancelled
    case .disconnected:
      return .disconnected
    case .operationAlreadyPending, .failed:
      return .failed
    }
  }
}
